### PR TITLE
DOC: Updating Chararry Buffer datatypes #15360

### PR DESCRIPTION
**Pull request by MLK97**
This pull request is motivated by the [this](https://github.com/numpy/numpy/issues/15360) documentation issue.
I added the str and bytes datatype, hopefully correct.

### CODE REVIEW

**Comment by MLK97** [comment]
I looked into the failed circleci test and can't really figure out where the error lies.
I build numpy and made the doc with make html and it doesn't show the change in the doc.
Could someone give me a hint, where I failed?

**Comment by seberg** [comment]
There was an issue in a different place that snuck in through some chance in the timeline of changes. Closing-reopening which will probably make it go away.

About the change itself: `int` is not actually supported anymore I believe (it seems it was python 2 only, and strange). I think it probably is either str or "object exposing the buffer interface" (identical to `np.ndarray` which has a `buffer` argument). But I am not quite sure, so if you can check that would be great.

File: numpy/core/defchararray.py [hunk 1]
```diff
@@ -1908,7 +1908,7 @@ class adds the following functionality:
     unicode : bool, optional
         Are the array elements of type unicode (True) or string (False).
         Default is False.
-    buffer : int, optional
+    buffer : int or str or bytes, optional
```
Comment (eric-wieser) [inline numpy/core/defchararray.py]: Is int actually allowed here?

Comment (seberg) [reply numpy/core/defchararray.py]: Without wanting to check, I somewhat expect this was that ominous `int_asbuffer` function that we removed in the p3k cleanup :).

**Comment by MLK97** [comment]
> There was an issue in a different place that snuck in through some chance in the timeline of changes. Closing-reopening which will probably make it go away.
> 
> About the change itself: `int` is not actually supported anymore I believe (it seems it was python 2 only, and strange). I think it probably is either str or "object exposing the buffer interface" (identical to `np.ndarray` which has a `buffer` argument). But I am not quite sure, so if you can check that would be great.

I looked up the function and you are right.
Even further the function itself makes use of `np.darray` here
`if buffer is None:
            self = ndarray.__new__(subtype, shape, (dtype, itemsize),
                                   order=order)
        else:
            self = ndarray.__new__(subtype, shape, (dtype, itemsize),
                                   buffer=buffer,
                                   offset=offset, strides=strides,
                                   order=order)`

So I will correct it to _object exposing buffer interface_
Should I also add str or just copy the way it is mentioned [here](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html)?

**Comment by seberg** [comment]
If you look at the code, unicode (python 3 just str), is handled differently though. It is used as a filler. Quite frankly, that is _not_ really good usage I think. But until we remove it, should be documented. So technically it is `...buffer interface... or str` right now.

I would love to say: Yes copy the one from `np.ndarray`... but it is wrong :). The correct spelling is probably more something like: `"Buffer which will be viewed as an array, so that the arrays views the same data as the buffer."`, unless there is a code path where we actually do a copy because the buffer is incompatible, hmmmm.

File: numpy/core/defchararray.py [hunk 2]
```diff
@@ -1908,7 +1908,7 @@ class adds the following functionality:
     unicode : bool, optional
         Are the array elements of type unicode (True) or string (False).
         Default is False.
-    buffer : int, optional
+    buffer : object exposing the buffer interface or str, optional
```
Comment (MLK97) [inline numpy/core/defchararray.py]: So I removed the int datatype and edited the file to fit your description.

File: numpy/core/defchararray.py (hunk 1 above)
Comment (mattip) [reply numpy/core/defchararray.py]: no, it is not allowed. `np.chararray((3, 3), buffer=12)` emits `TypeError: a bytes-like object is required, not 'int'`

**Comment by mattip** [comment]
In order to clear the error, you have to rebase off master:
```
git checkout master
git pull
git checkout DOC_correction_15360
git rebase master
<maybe fix errors, probably not>
git push -f MLK97
```

**Comment by mattip** [comment]
Something is off. I will fix it up and merge.

**Comment by mattip** [comment]
Thanks @MLK97

**Pull request merged**